### PR TITLE
fixes #6372 and #6094 - allowing to add org and loc to usergroups, using them in creation of external users

### PR DESCRIPTION
**Pull request by unorthodoxgeek**

### CODE REVIEW

File: app/views/usergroups/_form.html.erb [hunk 1]
```diff
@@ -58,6 +64,9 @@
         </div>
       </div>
     <% end %>
+
+    <%= render 'taxonomies/loc_org_tabs', :f => f, :obj => @usergroup
+      %>
```
Comment (dLobatog) [inline app/views/usergroups/_form.html.erb]: This can be closed in the line above

Comment (unorthodoxgeek) [inline app/views/usergroups/_form.html.erb]: yup. leftovers... fixing.

File: test/unit/user_test.rb [hunk 2]
```diff
@@ -552,6 +552,36 @@ def setup_user(operation)
         new_source = AuthSourceExternal.find_by_name('new_external_source')
         assert_equal new_source.name, created_user.auth_source.name
       end
+
+      context "usergroups and taxonomies association" do
+        before do
```
Comment (dLobatog) [inline test/unit/user_test.rb:571]: Would using `FactoryGirl.build` work? It would save us several hits to the db

Comment (unorthodoxgeek) [inline test/unit/user_test.rb:571]: no, the next lines depend on this existing in the db.

**Comment by thomasmckay** [comment]
Missing api controller updates? Need apipie docs and rabl changes, I think. Also, could you open an issue against hammer to update command there? Thanks!

**Comment by unorthodoxgeek** [comment]
@thomasmckay - I've looked at the API controllers and I don't see a reference to orgs/locations in any of the controllers... can you point me to where I can find a reference to what should be done?

**Comment by thomasmckay** [comment]
For rabl json example
https://github.com/theforeman/foreman/blob/develop/app/views/api/v2/users/main.json.rabl#L7

For the controller example
https://github.com/theforeman/foreman/blob/develop/app/controllers/api/v2/users_controller.rb#L10

**Comment by unorthodoxgeek** [comment]
@thomasmckay - I have updated the API controller (and the taxonomy edit UI)

**Comment by unorthodoxgeek** [comment]
[test]

**Comment by ohadlevy** [comment]
@eLobato ping?

File: app/controllers/api/v1/usergroups_controller.rb [hunk 3]
```diff
@@ -24,6 +24,7 @@ def show
       api :POST, "/usergroups/", "Create a user group."
       param :usergroup, Hash, :required => true do
         param :name, String, :required => true
+        param_group :taxonomies, ::Api::V2::BaseController
```
Comment (dLobatog) [inline app/controllers/api/v1/usergroups_controller.rb:27]: v2 base controller? I'd say don't bother with v1.

File: app/models/user.rb [hunk 4]
```diff
@@ -222,6 +222,18 @@ def post_successful_login
     User.current = self
   end
 
+  def set_locations_and_organizations_from_usergroups(usergroups)
+    new_locations = usergroups.map(&:locations).flatten - locations
```
Comment (dLobatog) [inline app/models/user.rb]: Use `Taxonomy.enabled_taxonomies` and do this in a loop or something, it's literally the same code for both taxonomies, and one or the other might not be enabled.

**Comment by dLobatog** [comment]
@unorthodoxgeek I'll get back to this tomorrow so feel free to not make changes until I finish up the review

**Comment by unorthodoxgeek** [comment]
@eLobato - can you re-review please?

File: app/views/taxonomies/_form.html.erb [hunk 5]
```diff
@@ -7,6 +7,9 @@
       <% if User.current.allowed_to?(:view_users) %>
         <li><a href="#users" data-toggle="tab"><%= _("Users") %></a></li>
       <% end %>
+      <% if User.current.allowed_to?(:view_usergroups) %>
+        <li><a href="#usergroups" data-toggle="tab"><%= _("Usergroups") %></a></li>
```
Comment (dLobatog) [inline app/views/taxonomies/_form.html.erb]: `User Groups` with a space, not sure about the titleization, I'd go for "User groups" but then again I see "Smart Proxies" below...

File: app/models/user.rb [hunk 6]
```diff
@@ -228,6 +228,15 @@ def post_successful_login
     User.current = self
   end
 
+  def set_locations_and_organizations_from_usergroups(usergroups)
+    Taxonomy.enabled_taxonomies.each do |taxonomy|
+      new_taxonomies = usergroups.map{|i| i.send(taxonomy)}.flatten - send(taxonomy)
```
Comment (dLobatog) [inline app/models/user.rb]: Add `self` to this send or remove it from the other self below, just want to keep the method consistent.

File: app/models/user.rb [hunk 7]
```diff
@@ -228,6 +228,15 @@ def post_successful_login
     User.current = self
   end
 
+  def set_locations_and_organizations_from_usergroups(usergroups)
```
Comment (dLobatog) [inline app/models/user.rb]: This should probably be private, the method doesn't get called from anywhere else. Also the naming, should it be `set_taxonomies` perhaps?

File: test/unit/user_test.rb [hunk 8]
```diff
@@ -552,6 +552,36 @@ def setup_user(operation)
...
+                                                     :name => @auth_source.name).save(:validate => false)
+          as_admin do
+            @loc = FactoryGirl.create(:location)
+            @org = FactoryGirl.create(:organization)
+          end
+          @usergroup.locations << @loc
+          @usergroup.organizations << @org
+        end
+
+        test "user gets assigned to the corrent usergroups" do
```
Comment (dLobatog) [inline test/unit/user_test.rb]: `corrent` -> `current`  , same for the following test

**Comment by dLobatog** [comment]
@unorthodoxgeek One thing aside from the in-line comments, `set_taxonomies_from_usergroups` gets called only on creation of external users, but as it is right now, I see a few problems:
1. Already created users in an user group don't get the taxonomies, as these are only added on creation.
2. Removing the user group, or taxonomies from the user group does not remove them from the users. I think this is quite confusing authorization-wise, and we should try to resolve taxonomies at runtime rather than plainly adding them to the User model. 
3. Non-external users in an user group are part of an user group in a taxonomy, yet they don't get it at any point. Is there a reason for this?

@ares I feel like explaining your point of view on 2 could help here, what would it be?

Thanks for the work so far, one step closer.

**Comment by ares** [comment]
If I understand correctly, taxonomies assigned to user groups are only used as a template set for newly created external user. Right now we don't know into which org we should assign the user. If that's the case, I think it's OK but I would prefer to have this association out of user group, ideally in ExternalUsergroup. Also orgs might be set by ENV varibales if we use mod_identity_lookup so maybe a bit more generic code would be nice (but it's fair to say it's out of scope).

If I'm wrong and this is a way how to deal with assigned taxonomies for all users through user groups then I agree with @eLobato, we should apply them in runtime (maybe use the same caching mechanism as we do for permissions) and we should disassociate taxonomies from user when they are disassociated from user groups. But I hope this isn't the case because this would "hardcode" taxonomies being used for authorization even more :-(

**Comment by unorthodoxgeek** [comment]
from what I understand, what matters is the creation, afterwards, we can edit the permissions on a per-user level. this fixes an annoying behaviour where users can't use foreman if they're new and logged in from LDAP.

@eLobato @ares correct me if I'm wrong...

p.s. updated according to Daniel's comments

**Comment by ares** [comment]
So it's just a template for taxonomies for new users. In that case would it be hard to move the taxonomies relation to ExternalUsergroup so it's more clear?

**Comment by dLobatog** [comment]
I don't think as it stands right now it could go in, some of the issues I'm finding are:
- Users who already have logged in simply don't get the taxonomies from its external user group, but users who are newly created (after the patch) do get the taxonomies.
- It introduces a security gap where:
  - User 'tom' gets Organizations 'foo' , 'bar' after getting a new account created as part of the External User Group 'baz'
  - On LDAP, the administrator decides to remove the group 'baz'.
  - 'tom' still has access to Organizations 'foo' and 'bar', yet he's not even member of the User Group anymore. This is never fixed unless an admin takes manual action.
- Having the Taxonomies tabs we get everywhere on the User group edit view has a very different meaning than in the rest of models. It makes me think Users in the User group will get these Taxonomies, which is a half-truth. I'd expect many people to report bugs on this if it was merged now.

I could even justify closing the bug if there was a workaround with roles, I tried giving an user with no permissions just `view_organizations, assign_organizations` but it doesn't seem to be enough

**Pull request closed without merging**